bigcache.go

```diff
@@ -33,6 +33,111 @@ type cacheShard struct {
 	onRemove    func(wrappedEntry []byte)
 }
 
+// EntryInfo holds informations about entry in the cache
+type EntryInfo struct {
+	entry []byte
+}
+
+// Key returns entry's underlying key
+func (e EntryInfo) Key() string {
+	return readKeyFromEntry(e.entry)
+}
+
+// Hash returns entry's hash value
+func (e EntryInfo) Hash() uint64 {
+	return readHashFromEntry(e.entry)
+}
+
+// Timestamp returns entry's timestamp (time of insertion)
+func (e EntryInfo) Timestamp() uint64 {
+	return readTimestampFromEntry(e.entry)
+}
+
+// Value returns entry's underlying value
+func (e EntryInfo) Value() []byte {
+	return readEntry(e.entry)
+}
+
+// EntryInfoIterator allows to iterate over entries in the cache
+type EntryInfoIterator struct {
+	cache        *BigCache
+	currentShard int
+	currentIndex int32
+	elements     []uint32
+	sync.Mutex
+}
+
+func copyCurrentShardMap(shard *cacheShard) []uint32 {
+	shard.lock.RLock()
+	defer shard.lock.RUnlock()
+
+	var elements []uint32
+
+	for _, index := range shard.hashmap {
+		elements = append(elements, index)
+	}
+
+	return elements
+}
+
+// Next returns true if there is next element in the iterator
+func (it *EntryInfoIterator) Next() bool {
+	it.Lock()
+	defer it.Unlock()
+
+	it.currentIndex++
+
+	if it.currentIndex >= int32(len(it.elements)) {
+	incrementShard:
+		it.currentIndex = 0
+		it.currentShard++
+
+		if it.currentShard == it.cache.config.Shards {
+			return false
+		}
+
+		it.elements = copyCurrentShardMap(it.cache.shards[it.currentShard])
+
+		// partition is empty - move to next
+		if len(it.elements) == 0 {
+			goto incrementShard
```

Comment:
Yep, that's exactly what I'm doing and code is readable for me. Goto is a standard jmp in the code